ssl/record/rec_layer_s3.c

```diff
@@ -1133,13 +1165,24 @@ int ssl3_write_pending(SSL *s, int type, const unsigned char *buf, size_t len,
         clear_sys_error();
         if (s->wbio != NULL) {
             s->rwstate = SSL_WRITING;
+            if (BIO_get_offload_tx(s->wbio)
+                && type != SSL3_RT_APPLICATION_DATA) {
+                BIO_set_offload_tx_ctrl_msg(s->wbio, type);
+            }
             /* TODO(size_t): Convert this call */
             i = BIO_write(s->wbio, (char *)
                           &(SSL3_BUFFER_get_buf(&wb[currbuf])
                             [SSL3_BUFFER_get_offset(&wb[currbuf])]),
                           (unsigned int)SSL3_BUFFER_get_left(&wb[currbuf]));
             if (i >= 0)
                 tmpwrit = i;
+
+            /*
+             * To prevent coalescing of control and data messages,
+             * such as in buffer_write, we flush the BIO
+             */
+            if (BIO_get_offload_tx(s->wbio) && type != SSL3_RT_APPLICATION_DATA)
+                (void)BIO_flush(s->wbio);
```

Comment:
This could fail!

Reply:
Are you concerned about non-blocking IO or some other failure?
If this is a ktls socket, then non-blocking IO sendmsg will either succeed fully or fail fully. There is no problematic case where the socket writes some partial payload.
In case of failure, we expect sock_write to be called again with the control message and any additional data of the same record type.

Reply:
I'm thinking about non-blocking IO mainly - but other failures are also a concern I supposed. Are you saying with ktls we will never have data pending in the BIO so we don't need to worry about this (in which case why are we flushing at all)?

Reply:
This comment seems to be outstanding

Reply:
In #7527 you said:
> What do you mean? - AFAIU it is resolved.

I still had a question on this which was unanswered, i.e. "Are you saying with ktls we will never have data pending in the BIO so we don't need to worry about this (in which case why are we flushing at all)?"

Reply:
Oh, I've missed this.

You're right this might fail and we should handle this, I'll add some code to handle this failure.